### ISSUE TITLE
APIv2: protocmp: crash comparing nils with embedded messages

### ISSUE THREAD

**Issue by neild** [golang/protobuf#1010]
[CL 213238](https://go-review.googlesource.com/c/protobuf/+/213238) introduces a panic when transforming a nil pointer to a type embedding a `proto.Message`. Maybe we're okay saying it's the responsibility of the caller to provide us with something whose `ProtoReflect` method doesn't panic.

```
--- a/testing/protocmp/xform_test.go
+++ b/testing/protocmp/xform_test.go
@@ -21,12 +21,18 @@ func init() {
        detrand.Disable()
 }

+type Embed struct {
+       *testpb.TestAllTypes
+}
+
 func TestTransform(t *testing.T) {
        tests := []struct {
                in         proto.Message
                want       Message
                wantString string
        }{{
+               in: (*Embed)(nil),
+       }, {
                in: &testpb.TestAllTypes{
                        OptionalBool:          proto.Bool(false),
                        OptionalInt32:         proto.Int32(-32),
```

**Comment by dsnet** [comment]
I've spent years arguing against embedding of protobuf messages, and this issue with nil derefencing was one of the arguments why it was a bad idea. 

When we added the `Message.IsValid` method, it was supposed to be the principled way that we detect invalid messages. If we can't make use of that method here, then I'm not entirely sure we should add `IsValid`. Personally, I think we should push back on these broken use cases.
